python-package/lightgbm/callback.py

```diff
@@ -209,12 +215,69 @@ def _init(env):
                 best_score.append(float('inf'))
                 cmp_op.append(lt)
 
+    def _metric_alias_matching(target_metric):
+        """Convert metric name for early stopping from alias to representative in order to match the param metric."""
+        metrics_list = [("l1", {"mean_absolute_error", "mae", "regression_l1"}),
+                        ("l2", {"mean_squared_error", "mse", "regression_l2", "regression"}),
+                        ("l2_root", {"root_mean_squared_error", "rmse"}),
+                        ("mape", {"mean_absolute_percentage_error"}),
+                        ("ndcg", {"lambdarank"}),
+                        ("map", {"mean_average_precision"}),
+                        ("binary_logloss", {"binary"}),
+                        ("multi_logloss",
+                        {"multiclass", "softmax", "multiclassova", "multiclass_ova", "ova", "ovr"}),
+                        ("xentropy", {"cross_entropy"}),
+                        ("xentlambda", {"cross_entropy_lambda"}),
+                        ("kldiv", {"kullback_leibler"})]
+        for metric, aliases in metrics_list:
+            if target_metric in aliases:
+                return metric
+        return target_metric
+
     def _callback(env):
         if not cmp_op:
             _init(env)
         if not enabled[0]:
             return
+        if first_metric_only:
+            # Choosing a target metric for early stopping from first element of metrics list as `eval_metric`
+            # if first_metric_only == True
+            eval_metric = None
+            for metric_alias in ['metric', 'metrics', 'metric_types']:
+                if metric_alias in env.params.keys():
+                    if isinstance(env.params[metric_alias], (tuple, list)):
+                        metric_list = [m for m in env.params[metric_alias] if m is not None]
+                        eval_metric = metric_list[0]
+                    else:  # string
+                        if env.evaluation_result_list[0][0] == "cv_agg":  # for lgb.cv
+                            for i in range(len(env.evaluation_result_list)):
+                                # Removing training data
+                                if ((env.evaluation_result_list[i][1].find(" ") >= 0
+                                     and env.evaluation_result_list[i][1].split(" ")[0] != "train")):
+                                    eval_metric = env.evaluation_result_list[i][1].split(" ")[1]
+                                    break
+                        else:  # the other than lgb.cv
+                            eval_metric = env.params[metric_alias]
+                    break
+            if eval_metric is None:
+                # If `eval_metric` was not chosen from `env.params`,
+                # using first element of env.evaluation_result_list for early stopping
+                eval_metric = env.evaluation_result_list[0][1]
+            eval_metric = _metric_alias_matching(eval_metric)
```

Comment:
@matsuken92  Seems that alias mapping doesn't work 😢 
Compare.

### `metric=['l2']` - OK:

```
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

import lightgbm as lgb

X_train, X_test, y_train, y_test = train_test_split(*load_boston(True), test_size=0.2, random_state=42)
lgb_train = lgb.Dataset(X_train, y_train)
lgb_valid = lgb.Dataset(X_test, y_test, reference=lgb_train)
lgb.train({'objective': 'regression', 'metric': ['l2'], 'first_metric_only': True, 'learning_rate': 1},
          lgb_train, num_boost_round=30, valid_sets=[lgb_valid], early_stopping_rounds=1)
```

```
[1]	valid_0's l2: 19.4941
Training until validation scores don't improve for 1 rounds.
[2]	valid_0's l2: 18.2572
[3]	valid_0's l2: 18.1671
[4]	valid_0's l2: 19.1842
Early stopping, best iteration is:
[3]	valid_0's l2: 18.1671
Evaluated only: l2
```

### `metric=['l1']` - OK:

```
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

import lightgbm as lgb

X_train, X_test, y_train, y_test = train_test_split(*load_boston(True), test_size=0.2, random_state=42)

lgb_train = lgb.Dataset(X_train, y_train)
lgb_valid = lgb.Dataset(X_test, y_test, reference=lgb_train)
lgb.train({'objective': 'regression', 'metric': ['l1'], 'first_metric_only': True, 'learning_rate': 1},
          lgb_train, num_boost_round=30, valid_sets=[lgb_valid], early_stopping_rounds=1)
```

```
[1]	valid_0's l1: 2.7825
Training until validation scores don't improve for 1 rounds.
[2]	valid_0's l1: 2.8229
Early stopping, best iteration is:
[1]	valid_0's l1: 2.7825
Evaluated only: l1
```

### `metric=['rmse']` - fail:

```
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

import lightgbm as lgb

X_train, X_test, y_train, y_test = train_test_split(*load_boston(True), test_size=0.2, random_state=42)
lgb_train = lgb.Dataset(X_train, y_train)
lgb_valid = lgb.Dataset(X_test, y_test, reference=lgb_train)
lgb.train({'objective': 'regression', 'metric': ['rmse'], 'first_metric_only': True, 'learning_rate': 1},
          lgb_train, num_boost_round=30, valid_sets=[lgb_valid], early_stopping_rounds=1)
```

```
[1]	valid_0's rmse: 4.41521
Training until validation scores don't improve for 1 rounds.
[2]	valid_0's rmse: 4.27284
[3]	valid_0's rmse: 4.26228
[4]	valid_0's rmse: 4.37998
[5]	valid_0's rmse: 4.10348
[6]	valid_0's rmse: 4.09483
[7]	valid_0's rmse: 3.99508
[8]	valid_0's rmse: 3.85972
[9]	valid_0's rmse: 3.84373
[10]	valid_0's rmse: 3.83278
[11]	valid_0's rmse: 3.8154
[12]	valid_0's rmse: 3.83956
[13]	valid_0's rmse: 3.81204
[14]	valid_0's rmse: 3.85466
[15]	valid_0's rmse: 3.85519
[16]	valid_0's rmse: 3.79894
[17]	valid_0's rmse: 3.7485
[18]	valid_0's rmse: 3.83072
[19]	valid_0's rmse: 3.78882
[20]	valid_0's rmse: 3.80427
[21]	valid_0's rmse: 3.80187
[22]	valid_0's rmse: 3.76381
[23]	valid_0's rmse: 3.76004
[24]	valid_0's rmse: 3.71417
[25]	valid_0's rmse: 3.76152
[26]	valid_0's rmse: 3.7486
[27]	valid_0's rmse: 3.77054
[28]	valid_0's rmse: 3.7859
[29]	valid_0's rmse: 3.80075
[30]	valid_0's rmse: 3.81967
```

### `metric=['l2_root']` - fail:


```
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

import lightgbm as lgb

X_train, X_test, y_train, y_test = train_test_split(*load_boston(True), test_size=0.2, random_state=42)
lgb_train = lgb.Dataset(X_train, y_train)
lgb_valid = lgb.Dataset(X_test, y_test, reference=lgb_train)
lgb.train({'objective': 'regression', 'metric': ['l2_root'], 'first_metric_only': True, 'learning_rate': 1},
          lgb_train, num_boost_round=30, valid_sets=[lgb_valid], early_stopping_rounds=1)
```

```
[1]	valid_0's rmse: 4.41521
Training until validation scores don't improve for 1 rounds.
[2]	valid_0's rmse: 4.27284
[3]	valid_0's rmse: 4.26228
[4]	valid_0's rmse: 4.37998
[5]	valid_0's rmse: 4.10348
[6]	valid_0's rmse: 4.09483
[7]	valid_0's rmse: 3.99508
[8]	valid_0's rmse: 3.85972
[9]	valid_0's rmse: 3.84373
[10]	valid_0's rmse: 3.83278
[11]	valid_0's rmse: 3.8154
[12]	valid_0's rmse: 3.83956
[13]	valid_0's rmse: 3.81204
[14]	valid_0's rmse: 3.85466
[15]	valid_0's rmse: 3.85519
[16]	valid_0's rmse: 3.79894
[17]	valid_0's rmse: 3.7485
[18]	valid_0's rmse: 3.83072
[19]	valid_0's rmse: 3.78882
[20]	valid_0's rmse: 3.80427
[21]	valid_0's rmse: 3.80187
[22]	valid_0's rmse: 3.76381
[23]	valid_0's rmse: 3.76004
[24]	valid_0's rmse: 3.71417
[25]	valid_0's rmse: 3.76152
[26]	valid_0's rmse: 3.7486
[27]	valid_0's rmse: 3.77054
[28]	valid_0's rmse: 3.7859
[29]	valid_0's rmse: 3.80075
[30]	valid_0's rmse: 3.81967
```

I guess that this is due to that cpp code returns alias instead of main name in some cases (`rmse` instead of documented `l2_root`).

Reply:
Fixed in #2257 and 4221b8a657c0a23c2710a5a197b10b9fd67a17d3.